File: perf/benchmarks/micro/bit_count/bit_count.cpp
// bit_count — popcount tight-loop micro-benchmark (C++).
#include <chrono>
#include <cstdint>
#include <cstdio>
#include <cstdlib>

int main(int argc, char** argv) {
    int64_t n = (argc > 1) ? std::atoll(argv[1]) : 5000000;
    auto t0 = std::chrono::steady_clock::now();
    int64_t cs = 0;
    for (int64_t i = 0; i < n; ++i) {
        uint32_t v = static_cast<uint32_t>(i * 2654435761ULL);
        // Manual popcount to keep all five implementations apples-to-apples.
        uint32_t x = v;
        int c = 0;
        while (x) { x &= x - 1; ++c; }
        cs += c;
    }
    auto t1 = std::chrono::steady_clock::now();
    double ms = std::chrono::duration<double, std::milli>(t1 - t0).count();
    std::printf("CHECKSUM:%lld\n", static_cast<long long>(cs));
    std::printf("ELAPSED_MS:%.6f\n", ms);
    return 0;
}

File: perf/benchmarks/micro/bit_count/bit_count.py
# bit_count — popcount tight-loop micro-benchmark.
#
# For i in [0, N) compute popcount(i * 2654435761 & 0xFFFFFFFF) and sum.
# Pure integer / bitwise work. Argument: N (default 5_000_000).

import sys
import time


def popcount(x: int) -> int:
    c = 0
    while x:
        x &= x - 1
        c += 1
    return c


def main() -> None:
    n = int(sys.argv[1]) if len(sys.argv) > 1 else 5_000_000
    t0 = time.perf_counter_ns()
    cs = 0
    for i in range(n):
        v = (i * 2654435761) & 0xFFFFFFFF
        cs += popcount(v)
    elapsed_ns = time.perf_counter_ns() - t0
    print(f"CHECKSUM:{cs}")
    print(f"ELAPSED_MS:{elapsed_ns / 1_000_000:.6f}")


if __name__ == "__main__":
    main()
